### ISSUE TITLE
try/except usage of importlib.metadata/importlib_metadata does not work on 3.7 or lower

### ISSUE THREAD

**Issue by agp498** [Nuitka/Nuitka#1745]
Hello

I used nuitka 0.7.7 for a long time without problems.
After upgrade to 1.0.3 I started to receive an error (importlib_metadata.PackageNotFoundError) in my project that previously worked correct.

```
importlib_metadata.PackageNotFoundError: No package metadata was found for prettytable
```

### Nuitka version:
```
1.0.3
Commercial: None
Python: 3.6.8 (default, Aug 13 2020, 07:46:32)
Flavor: Fedora Python
Executable: /home/rlm_dev/ushd/sds_discovery/bin/python
OS: Linux
Arch: x86_64
Distribution: Rhel (based on Fedora) 7.9
```
 
### Pip list:
```
Package            Version
------------------ ---------
attrs              21.4.0
boto3              1.23.10
botocore           1.26.10
certifi            2022.6.15
charset-normalizer 2.0.12
idna               3.3
importlib-metadata 4.8.3
jmespath           0.10.0
Nuitka             1.0.3
orderedset         2.0.3
pip                21.3.1
prettytable        2.5.0
psycopg2-binary    2.9.3
python-dateutil    2.8.2
requests           2.27.1
requests-aws       0.1.8
rgwadmin           2.3.1
s3transfer         0.5.2
setuptools         39.2.0
six                1.16.0
typing_extensions  4.1.1
urllib3            1.26.9
wcwidth            0.2.5
zipp               3.6.0
```

I tested some versions of nuitka and found out that last worked version is 0.8.4 and problem started from 0.9.
Test program test.py:
 
```
import prettytable

def main():
    print("hello world")
    return

 if __name__ == '__main__':
    main()
```
 
Nuitka 1.0.3:
 ```
(sds_discovery) [[rlm_dev@tkles](mailto:rlm_dev@tkles) /home/rlm_dev/ushd/sds_discovery/sds/sds/sds_discovering]$ python -m nuitka --standalone test.py
Nuitka-Options:INFO: Used command line options: --standalone test.py
Nuitka:INFO: Starting Python compilation with Nuitka '1.0.3' on Python '3.6' commercial grade 'not installed'.
Nuitka-Plugins:INFO: multiprocessing: Injecting pre-module load code for module 'multiprocessing':
Nuitka-Plugins:INFO: multiprocessing:     Monkey patching "multiprocessing" load environment.
Nuitka-Plugins:INFO: multiprocessing: Injecting post-module load code for module 'multiprocessing':
Nuitka-Plugins:INFO: multiprocessing:     Monkey patching "multiprocessing" for compiled methods.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C compilation via Scons.
Nuitka-Scons:INFO: The provided gcc is too old, switching to its g++ instead.
Nuitka-Scons:INFO: Backend C compiler: g++ (g++).
Nuitka-Scons:INFO: Backend linking program with 22 files (no progress information available).
Nuitka-Scons:INFO: Compiled 22 C files using ccache.
Nuitka-Scons:INFO: Cached C files (using ccache) with result 'cache hit': 6
Nuitka-Scons:INFO: Cached C files (using ccache) with result 'cache miss': 15
Nuitka:INFO: Keeping build directory 'test.build'.
Nuitka:INFO: Successfully created 'test.dist/test'.

 Traceback (most recent call last):
  File "/home/rlm_dev/ushd/sds_discovery/sds/sds/sds_discovering/test.dist/test.py", line 1, in <module>
  File "/home/rlm_dev/ushd/sds_discovery/sds/sds/sds_discovering/test.dist/prettytable/__init__.py", line 52, in <module prettytable>
  File "/home/rlm_dev/ushd/sds_discovery/sds/sds/sds_discovering/test.dist/importlib_metadata/__init__.py", line 990, in version
  File "/home/rlm_dev/ushd/sds_discovery/sds/sds/sds_discovering/test.dist/importlib_metadata/__init__.py", line 963, in distribution
  File "/home/rlm_dev/ushd/sds_discovery/sds/sds/sds_discovering/test.dist/importlib_metadata/__init__.py", line 542, in from_name
importlib_metadata.PackageNotFoundError: No package metadata was found for prettytable
```

 Nuitka 0.8.4:
 ```
INFO:  Start nuitka
Nuitka-Options:INFO: Used command line options: --standalone sds/sds_discovering/scripts/sds_discovery.py
Nuitka:INFO: Starting Python compilation with Nuitka '0.8.4' on Python '3.6' commercial None.
Nuitka-Plugins:INFO: anti-bloat: Handling module 'certifi.core' with 1 change(s) for: avoid using importlib.resources without need.
Nuitka-Plugins:INFO: multiprocessing: Injecting pre-module load code for module 'multiprocessing':
Nuitka-Plugins:INFO: multiprocessing:     Monkey patching "multiprocessing" load environment.
Nuitka-Plugins:INFO: multiprocessing: Injecting post-module load code for module 'multiprocessing':
Nuitka-Plugins:INFO: multiprocessing:     Monkey patching "multiprocessing" for compiled methods.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'urllib3' added 'urllib3,urllib3._collections,urllib3.connection,urllib3.connectionpool,urllib3.contrib,urllib3.contrib.appengine,urllib3.exceptions,urllib3.fields,urllib3.filepost,urllib3.packages,urllib3.packages.six,urllib3.poolmanager,urllib3.request,urllib3.response,urllib3.util,urllib3.util.connection,urllib3.util.queue,urllib3.util.request,urllib3.util.response,urllib3.util.retry,urllib3.util.ssl_,urllib3.util.timeout,urllib3.util.url,urllib3.util.wait,urllib.error,urllib.parse,urllib.request,urllib.response'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'boto3' added 'boto3.ec2,boto3.ec2.createtags,boto3.ec2.deletetags,boto3.dynamodb,boto3.s3,boto3.s3.inject,boto3.s3.transfer'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'zipfile' added 'py_compile'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'urllib.request' added 'http.client,http.cookiejar'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'socket' added '_socket'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'shelve' added 'dbm.dumb'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'pathlib' added 'ntpath,posixpath'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'ctypes' added '_ctypes'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module '_asyncio' added 'asyncio'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'locale' added '_bootlocale'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'encodings' added 'encodings.idna,encodings.base64_codec,encodings.hex_codec,encodings.rot_13'.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C compilation via Scons.
Nuitka-Scons:INFO: The provided gcc is too old, switching to its g++ instead.
Nuitka-Scons:INFO: Backend C compiler: g++ (g++).
Nuitka-Scons:INFO: Backend linking program (no progress information available).
Nuitka-Scons:INFO: Compiled 273 C files using ccache.
Nuitka-Scons:INFO: Cached C files (using ccache) with result 'cache hit': 272
Nuitka-Plugins:INFO: data-files: Included 16 data files due to package data directory 'data' for 'boto3'.
Nuitka-Plugins:INFO: data-files: Included data file 'botocore/cacert.pem' due to package data for 'botocore'.
Nuitka-Plugins:INFO: data-files: Included 906 data files due to package data directory 'data' for 'botocore'.
Nuitka-Plugins:INFO: data-files: Included data file 'certifi/cacert.pem' due to package data for 'certifi'.
Nuitka:INFO: Keeping build directory 'sds_discovery.build'.
Nuitka:INFO: Successfully created 'sds_discovery.dist/sds_discovery'.

(sds_discovery) [[rlm_dev@tkles](mailto:rlm_dev@tkles) /home/rlm_dev/ushd/sds_discovery/sds/sds/sds_discovering]$ ./test.dist/test
hello world
```
 
In addition I tried some options for solving the problem. Without success.
 ```
python -m nuitka --standalone --include-package=prettytable test.py
python -m nuitka --standalone --follow-stdlib test.py
python -m nuitka --standalone --full-compat test.py
python -m nuitka --standalone --disable-plugin=anti-bloat test.py
```

Could you help me to solve this problem?

Thanks.

**Comment by kayhayen** [comment]
Are you using prettytable from the RPM or from PyPI?

**Comment by agp498** [comment]
> Are you using prettytable from the RPM or from PyPI?

PyPI

**Comment by kayhayen** [comment]
I managed to reproduce this, my mistake was to use 3.9 and not 3.6 as you do.

**Comment by kayhayen** [comment]
It seems, this pattern is not working for static optimization in Nuitka:

````
try:
    # Python 3.8+
    import importlib.metadata as importlib_metadata
except ImportError:
    # <Python 3.7 and lower
    import importlib_metadata

__version__ = importlib_metadata.version(__name__)
```

I think for 3.8+ the exception handler goes away, and the hard import is being done and compile time resolves it to the version value. When the import is not available at compile time, I think so far it is not statically converted to an `ImportError` not is that exception handler optimized away. I will have to check it out.

**Comment by chrisy** [comment]
Possibly related (let me know if a new issue is required), I was chasing what I think is a related issue with `jsonschema`. They switched to `importlib` a while ago
(https://github.com/python-jsonschema/jsonschema/commit/ffb5e06ec6683592f89e3fef5888a3dd74975110) and when we bumped versions to include that change in our build, Nuitka doesn't seem to be doing the right thing. Pinning `jsonschema` to a version before this change (4.1.0) recovers the build.

Symptoms: importing `jsonschema` fails with a trace that ends with:

```
  File "/.../xyz.dist/jsonschema/_utils.py", line 61, in load_schema
  File "/.../xyz.dist/importlib/abc.py", line 378, in read_text
  File "/.../xyz.dist/importlib/_adapters.py", line 54, in open
ValueError
```

I did some limited diagnosis, and found that the `load_schema` method has the line `path = resources.files(__package__).joinpath(f"schemas/{name}.json")` which ends up with the object `<importlib._adapters.DegenerateFiles.Path object at 0x7f933d5a91b0>`. That `DegenerateFiles.Path` object seems to be a placeholder for resources without a reader.

My environment is Linux on Python 3.10 building a standalone application. I've been using the tip of `develop` of late, but based on the above commentary I suspect release versions have the same issue. I was able to reproduce the issue with this simple script:

```python
import jsonschema

if __name__ == "__main__":
    print("Hello!")
```

**Comment by kayhayen** [comment]
@chrisy definitely something that is a different issue

**Comment by kayhayen** [comment]
The absent imports from the first import leg now give static `ImportError` and therefore are no longer assignments, giving the other branch a chance to become the main branch, however this is not yet an optimization being done.

**Comment by kayhayen** [comment]
When I try either orientation of the imports in the try/except with the backport installed, and stdlib having `importlib.metadata` it works, but it doesn't work when the backport is tried first, but not installed, the issue with the exception not leading to optimization kicks in, which is kind of ugly. This must be an issue with the merging of branches in the try node, otherwise there would only be a single assignment, that should be used.

**Comment by kayhayen** [comment]
Actually, the only thing that was missing is for the list of trusted imports, that are creating an exception, to not only do `importlib.metadata` but also `importlib_metadata` and then it's just fine with the example code. The one that succeeds if both would, removes the exception handler, and therefore it works.

**Comment by kayhayen** [comment]
Thanks for your report, this is fixed on the factory branch, which is a development version under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 

Feedback if this is working is very welcome, just please do not share plans of doing it, but rather confirmations or denials of it working. 

Also, this will be in 1.2 release, which is coming shortly.

**Comment by kayhayen** [comment]
This is now on the develop branch, part of the current pre-release and should be part of the immanent 1.2 release.

I found that indeed, for some variations, there was a lack of static optimization inside the exception handler. Tried blocks that do not exit, still contributed to the trace afterwards in the form of a merge with an exiting exception handler part. That is not useful. Inside the exception handler, the re-raise was also leading to a merge of the other branch with its start point, that is equally useless. And then the check against a known exception for match was also annotating a control flow escape, that was useless. This all added, should mean all variations on all versions with all install states work now.
